crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,764 @@
-use std::borrow::Cow;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
 
-use runix::installable::{FlakeAttribute, ParseInstallableError};
-use serde_json::Value;
+use flox_types::catalog::{CatalogEntry, EnvCatalog};
+use log::debug;
+use rnix::ast::{AttrSet, Expr};
+use rowan::ast::AstNode;
+use runix::arguments::eval::EvaluationArgs;
+use runix::arguments::{BuildArgs, EvalArgs};
+use runix::command::{Build, Eval};
+use runix::command_line::{NixCommandLine, NixCommandLineRunError, NixCommandLineRunJsonError};
+use runix::flake_ref::path::PathRef;
+use runix::installable::FlakeAttribute;
+use runix::store_path::StorePath;
+use runix::{Run, RunJson};
 use thiserror::Error;
+use walkdir::WalkDir;
 
-use super::flox_package::FloxPackage;
-use super::floxmeta::environment::GenerationError;
-use super::root::transaction::ReadOnly;
-use super::{floxmeta, project};
-use crate::providers::git::GitProvider;
+use super::environment_ref::{EnvironmentName, EnvironmentRef, EnvironmentRefError};
+use super::flox_package::{FloxPackage, FloxTriple};
+use crate::utils::copy_file_without_permissions;
+use crate::utils::rnix::{AttrSetExt, StrExt};
 
 pub static CATALOG_JSON: &str = "catalog.json";
 // don't forget to update the man page
 pub const DEFAULT_KEEP_GENERATIONS: usize = 10;
 // don't forget to update the man page
 pub const DEFAULT_MAX_AGE_DAYS: u32 = 90;
 
-pub enum CommonEnvironment<'flox, Git: GitProvider> {
-    Named(floxmeta::environment::Environment<'flox, Git, ReadOnly<Git>>),
-    Project(project::environment::Environment<'flox, Git, ReadOnly<Git>>),
+pub enum InstalledPackage {
+    Catalog(FloxTriple, CatalogEntry),
+    FlakeAttribute(FlakeAttribute, CatalogEntry),
+    StorePath(StorePath),
 }
 
-impl<'flox, Git: GitProvider> CommonEnvironment<'flox, Git> {
-    /// get a flake attribute for the environment
-    /// todo flake_attribute should be constructed earlier
-    pub async fn flake_attribute(
-        &self,
-    ) -> Result<FlakeAttribute, EnvironmentError<GenerationError<Git>, ParseInstallableError>> {
-        match self {
-            CommonEnvironment::Named(n) => n
-                .flake_attribute(Default::default())
-                .await
-                .map_err(EnvironmentError::Named),
-            CommonEnvironment::Project(p) => p.flake_attribute().map_err(EnvironmentError::Project),
-        }
+/// Struct representing a local environment in a given location and state
+#[derive(Debug)]
+pub struct Environment<S> {
+    /// absolute path to the environment, typically within `<...>/.flox/name`
+    path: PathBuf,
+    /// Access state of the environment
+    ///
+    /// Implementations distinguish whether whe can [Modify] or only [Read] an environment
+    state: S,
```

Comment:
I think we should clarify what it means to "modify" the environment. Do we only include modifications to the declarative definition of the environment, or do we also include operations that have side effects e.g. `build`? For example, `build` doesn't change the environment definition, but it has side effects in that it overwrites `result`.